cmd/jag/commands/run.go

```diff
@@ -60,17 +104,72 @@ func RunCmd() *cobra.Command {
 			if err != nil {
 				return err
 			}
-			snapshot := filepath.Join(snapshotsCache, device.ID+".snapshot")
 
-			err = sdk.Compile(ctx, snapshot, entrypoint)
+			var snapshot string = ""
+
+			if strings.HasSuffix(entrypoint, ".snapshot") || strings.HasSuffix(entrypoint, ".snap") {
+				snapshot = entrypoint
+			} else {
+				// We are running a toit file, so we need to compile it to a
+				// snapshot first.
+				tempdir, err := ioutil.TempDir("", "jag_run")
+				if err != nil {
+					return err
+				}
+				defer os.RemoveAll(tempdir)
+
+				snapshotFile, err := ioutil.TempFile(tempdir, "jag_snapshot")
+				if err != nil {
+					return err
+				}
+				snapshot = snapshotFile.Name()
+				err = sdk.Compile(ctx, snapshot, entrypoint)
+				if err != nil {
+					// We assume the error has been printed.
+					// Mark the command as silent to avoid printing the error twice.
+					cmd.SilenceErrors = true
+					return err
+				}
+			}
+
+			programId, err := GetUuid(snapshot)
 			if err != nil {
-				// We assume the error has been printed.
-				// Mark the command as silent to avoid printing the error twice.
-				cmd.SilenceErrors = true
 				return err
 			}
 
-			b, err := sdk.Build(ctx, device, snapshot)
+			cacheSnapshot := filepath.Join(snapshotsCache, programId.String()+".snapshot")
+
+			// Copy the snapshot into the cache dir so it is available for
+			// decoding stack traces etc.
+			if cacheSnapshot != snapshot {
+				dest, err := ioutil.TempFile(snapshotsCache, "jag_snapshot")
+				if err != nil {
+					fmt.Printf("Failed to write temporary file in '%s'\n", snapshotsCache)
+					return err
+				}
+				defer dest.Close()
```

Comment:
I get confused about the patterns you use for closing? Do you actually need to close dest - twice? You defer close source and eagerly close it too?